api/src/services/user.service.ts

```diff
@@ -180,53 +182,85 @@ export class UserService {
   async authenticate({ grantType, email, password }: { grantType: GrantType; email: string; password?: string }): Promise<User> {
     const normalizedEmail = normalizeEmail(email);
     const user = await this.userRepo.findOneBy({ email: normalizedEmail });
+
     if (!user) {
       throw new UnauthorizedException('invalid credentials');
     }
 
+    const userKey = `user-${user.id}`;
     const timeThreshold = moment().subtract(15, 'minutes').toDate();
 
-    // If lockout time has passed, reset counter
+    let loginAttempts = await this.getLoginAttempts(userKey, user);
+
+    // Reset login attempts if the last login is older than 15 minutes
     if (user.lastLogin < timeThreshold) {
-      user.loginAttempts = 0;
-      // Otherwise abort request
-    } else if (user.loginAttempts >= 3) {
-      throw new TooManyRequestsException('too many login attempts');
+      loginAttempts = 0;
     }
 
-    switch (grantType) {
-      case GrantType.Password:
-        if (!user.encryptedPassword) {
-          await this.userRepo.increment({ id: user.id }, 'loginAttempts', 1);
-          throw new UnprocessableEntityException('No password for this user, was this account created via a provider?');
-        }
+    // Handle too many login attempts
+    if (loginAttempts >= 3) {
+      user.lastLogin = new Date();
+      await this.saveUser(user); // Save the updated lastLogin
```

Comment:
@samuelmbabhazi why we did not run `await this.userRepo.increment({ id: user.id }, 'loginAttempts', 1);` here? I would prefer to increase attempt counter in database in ANY case!